### PR TITLE
Move VkScript VertexData to a Buffer

### PR DESCRIPTION
**Pull request by dj2**
This CL uses the shared Buffer to hold the contents of the VertexData
block in VkScript instead of the ad-hoc Cell structure.

### CODE REVIEW

File: src/vkscript/parser.cc [hunk 1]
```diff
@@ -372,14 +375,19 @@ Result Parser::ProcessVertexDataBlock(const std::string& data) {
...
       }
     }
-    node->AddRow(std::move(row));
   }
 
-  node->SetHeaders(std::move(headers));
+  auto node = MakeUnique<VertexDataNode>();
+  for (size_t i = 0; i < headers.size(); ++i) {
+    auto buffer = MakeUnique<Buffer>(BufferType::kVertex);
+    buffer->SetData(std::move(values[i]));
```
Comment (dneto0) [inline src/vkscript/parser.cc]: The Buffer's DatumType  and Size attributes are not set.  Is that important?

File: src/vkscript/parser_test.cc [hunk 2]
```diff
@@ -312,17 +312,19 @@ TEST_F(VkScriptParserTest, VertexDataHeaderFormatString) {
...
-  EXPECT_EQ(1U, headers[1].location);
+  auto& header_1 = data->GetHeader(1);
+  EXPECT_EQ(1U, header_1.location);
   EXPECT_EQ(FormatType::kA8B8G8R8_UNORM_PACK32,
-            headers[1].format->GetFormatType());
+            header_1.format->GetFormatType());
+  EXPECT_TRUE(data->GetSegment(1).empty());
 }
 
 TEST_F(VkScriptParserTest, VertexDataHeaderGlslString) {
```
Comment (dneto0) [inline src/vkscript/parser_test.cc]: Not for this CL, but I'm confused by the  1/int/vec3 attribute header declaration. Is it sane to have a mismatch between the gl_type and the component of the glsl_type ?

Comment (dj2) [reply src/vkscript/parser_test.cc]: I have no idea, this is directly from VkRunner, so I have to replicate what they do to maintain compatibility.

File: src/buffer.h [hunk 3]
```diff
@@ -39,8 +39,12 @@ class Buffer {
   const DatumType& GetDatumType() const { return datum_type_; }
 
   void SetSize(size_t size) { size_ = size; }
-  size_t GetSize() const { return size_; }
-  size_t GetSizeInBytes() const { return size_ * datum_type_.SizeInBytes(); }
+  size_t GetSize() const {
+    if (size_ == 0 && !data_.empty())
```
Comment (dneto0) [inline src/buffer.h]: The test for data_.empty() is redundant. :-)

File: src/vkscript/parser.cc (hunk 1 above)
Comment (dneto0) [reply src/vkscript/parser.cc]: I think Buffer needs comments to describe its own internal data,on the data members and accessors.  The particularly odd one is the datum_type_ which should be described as only valid for an index buffer.

Comment (dj2) [reply src/vkscript/parser.cc]: Added a couple comments. I added the invalid DatumType comment in the VertexData node class as that's where it's invalid. Buffer is also used in the Amber code where the DatumType is valid. It's just with VertexData that it goes wonky due to things like form D24_UINT_8_SINT or D32_SFLOAT_8_UINT not really being representable.....

File: src/buffer.h (hunk 3 above)
Comment (dj2) [reply src/buffer.h]: Yea, removed.

**Comment by dj2** [comment]
I'm going to do a followup to this cl where I create:

```
class Buffer {
  std::vector<Value>
}

class FormatBuffer : public Buffer {
  Format format;
}

class DataBuffer: public Buffer {
  DatumType datum_type;
}
```

Then, the AmberScript buffers will be DataBuffers, the Index buffer will be a DataBuffer and the VertexData buffer will be a FormatBuffer. This will remove the weird restriction that the DatumType is wrong for VertexBuffer.

**Comment by dj2** [comment]
Actually, I'm going to close this and do it in other steps ....

**Pull request closed without merging**